### ISSUE TITLE
M.A.C.S. Basic Rifle Marksmanship 

### ISSUE THREAD

**Issue by hossbags2** [snes9xgit/snes9x#303]
Rom just flickers. If it can be fixed, please add support for Superscope. The light gun used for M.A.C.S was the Replica Jäger AP 74.

**Comment by bearoso** [comment]
This is probably a hard one.

**Comment by kps501** [comment]
With help, hacked snes9x build to play this game. Used Superscope as test replacement for MACS rifle (port 2). Played 3 shooting courses with high marks earned. Some notes since this is purely test hack.

static void DoGunLatch(int x, int y){
    PPU.GunVLatch = (uint16) (y + 42);
    PPU.GunHLatch = (uint16) (x + 36);

This gives perfect aim, according to in-game calibration. Which is very precise and picky.
(press joypad B to enter menu)


void S9xSetJoypadLatch(bool latch){ = left this alone actually

DoGunLatch(superscope.x, superscope.y);
- Unfortunately this needs to be called every frame. Game latches light pen constantly. Shows pink for offscreen (I think) and messes up your shots if not done correctly. Holding RMB and then firing registers shots.


Game constantly reads 4017 without latching most of time. This gave us problems and we don't know how to safely hack snes9x. We just did this which breaks joypad d-pad movement (menu). Another build had this removed which let us use menus. Lots of savestates to navigate menus.
		  case 0x4017:
                    S9xSetJoypadLatch(0);
                    S9xSetJoypadLatch(1);
                    S9xSetJoypadLatch(0);
                    return S9xReadJOYSERn(Address);


Finally we did this:
 	case SUPERSCOPE:
	if( superscope.read_buttons & 0x80 )
	return 0x01;
	else
	return 0x00;

	if (read_idx[n][0] < 8)
	return (bits | ((superscope.read_buttons & (0x80 >> read_idx[n][0]++)) ? 1 : 0));
	else
	{
	read_idx[n][0]++;
	return (bits | 1);
	}

Game only reads bit 0 of 4017. 1 = trigger. (see no$cash)


That about covers it other than this game looks playable! Hope this of some help to your team!

**Comment by bearoso** [comment]
I am totally unfamiliar with the Super Scope and related peripherals having never owned or wanted one. I think we'd have to special-case this peripheral in order to not break the Super Scope. I don't even have the super scope properly implemented for the GTK+ port, so that would be a little more work for me. :-)

**Comment by kps501** [comment]
Learned how to add a M.A.C.S. Rifle controller to controls.cpp. Internal testing is looking good so far. Going to put it through more prelim checks on a fork first before sending PR (enough changes).

**Comment by kps501** [comment]
Have a test build ready. Please try this:
https://ci.appveyor.com/api/buildjobs/cajwjc02o6pf6pe3/artifacts/snes9x-1.55-340-g1251cfe-win32.zip

Select Rifle controller before starting shooting simulator. There's a few minor code changes I want to make and tiny other things to check.

**Comment by hossbags2** [comment]
seems to work fine.....any way to stop the flickering when it first loads up before selecting rifle

**Comment by kps501** [comment]
Uhm. I know it's joypad 2 related - keeps strobing trigger on/off and blasts through the menus. Maybe snes9x devs have some approved ideas? I can only think of doing internal cartridge header check "MAC:Basic Rifle" and auto-selecting rifle only at bootup. This controller keeps updating values without needing latching.

I just select Rifle before starting rom. But I guess that isn't ideal either.

**Comment by bearoso** [comment]
> I just select manually Rifle before starting rom. 
> I can only think of doing internal cartridge header check "MAC:Basic Rifle" and auto-selecting rifle only at bootup. 

Those are basically your options.

**Comment by kps501** [comment]
PR test build ready.
https://ci.appveyor.com/api/buildjobs/cb0h7r32mhl2sfo9/artifacts/snes9x-1.55-352-g55c43ec-win32.zip
https://ci.appveyor.com/api/buildjobs/jj24sg01626n41wi/artifacts/snes9x-1.55-352-g55c43ec-win32-x64.zip

You can preview changes at github.com/kps501/snes9x

**Comment by hossbags2** [comment]
I like that....works better

**Comment by kps501** [comment]
I'll let devs comment before officially sending PR.

**Comment by bearoso** [comment]
I don't see anything offensive. I don't think "Rifle" everywhere in the source is very descriptive, but it should be OK. Since this mostly affects the Windows port, OV2 needs to review it.

**Comment by bearoso** [comment]
I don't know what those "Added files via upload" lines are about. They're definitely not useful commit history. Just diff the source tree all at once and make it one commit.

**Comment by kps501** [comment]
Thanks. I can change it to macsrifle - that should help. :)

**Comment by kps501** [comment]
Okay! Finally figured out how to squash, merge, diff everything. That's a neat trick.

Put new PR here - 1 clean commit
https://github.com/kps501/snes9x/tree/macsrifle

**Comment by OV2** [comment]
The win changes look good. The rest also looks fine, the only thing I could see is that it's missing saving the new internal fields to snapshots. I'm sure we can add that later if you make a PR.

**Comment by bearoso** [comment]
His earlier tree did save those to the snapshots. He must have missed that in the squash.

**Comment by kps501** [comment]
I guess so. Sent over the PR because don't want to accidentally muck something up again.

**Comment by bearoso** [comment]
Ok, merged that. Just create another PR with the snapshot changes in it. If you're having too much trouble with git, you can just post those files here and we'll commit them.

**Comment by kps501** [comment]
Hmm. As far I can tell, it's in snes9x master.

controls.cpp
S9xControlPreSaveState - uses COPY macro
S9xControlPostSaveState - uses other COPY macro

controls.h
SControlSnapshot = updated internal size from 60 -> 65 bytes for macsrifle

Not sure what I'm missing yet... ^^

**Comment by bearoso** [comment]
Yeah, that’s what I saw earlier. OV2 must’ve overlooked it, and I must’ve glazed over it on my second look. Sorry about that.

**Comment by OV2** [comment]
What I meant was the actual saving to the snapshot, which is controlled by the SnapControls structure in snapshot.cpp. I've pushed a fix that should work.

**Comment by kps501** [comment]
Oh! Okay! I get it now. I completely would've missed that fix for like forever. And I like how you separated v3 and v4 save checking.

**Comment by retrotalker** [comment]
Oggom mentions this flicker bug when joypad2 is plugged in. Inside controls.cpp is this

S9xReadJOYSERn
          case JOYPAD0: case JOYPAD1: case JOYPAD2: case JOYPAD3:
          case JOYPAD4: case JOYPAD5: case JOYPAD6: case JOYPAD7:
            if(read_idx[n][0]>=16){
                read_idx[n][0]++;
                return bits|1;
            } else {
                return bits|((joypad[i-JOYPAD0].buttons&(0x8000>>read_idx[n][0]++))?1:0);
            }


This game does manual polling and never resets read_idx[1][0]. Eventually it reads so many times it overflows read_idx counter 255 to 0 and starts reading joypad bits again. This triggers 0|1 pulses and game thinks rifle trigger is rapidly being pulled.

I don't know how this should be fixed though.

**Comment by OV2** [comment]
We can limit increases to read_idx to stop at 255. I'll try to add that.

**Comment by OV2** [comment]
I've pushed a change that should stop the overflow.

**Comment by retrotalker** [comment]
It works! No more flicker crazyness.

**Comment by bearoso** [comment]
Closing. If more issues appear, reopen.
